modules/derivation/shared/src/main/scala/io/circe/derivation/DerivationMacros.scala

```diff
@@ -738,10 +753,12 @@ class DerivationMacros(val c: blackbox.Context) extends ScalaVersionCompat {
         }
       case _root_.scala.Some(typeFieldName) =>
         new _root_.io.circe.Codec.AsObject[$tpe] {
-          private[this] val decoder =
-            _root_.io.circe.Decoder[_root_.java.lang.String].prepare(_.downField(typeFieldName)).flatMap {
-              case ..$discriminatorCases
-            }
+          private[this] val decoder = new _root_.io.circe.Decoder[$tpe] {
+            def apply(c: _root_.io.circe.HCursor): _root_.io.circe.Decoder.Result[$tpe] =
+              c.get(typeFieldName)(_root_.io.circe.Decoder[_root_.java.lang.String]).flatMap {
+                case ..$discriminatorCases
```

Comment:
Other question I remembered: This code path didn't have a catch-all `case _ => ???` like the object-wrapper case did. Is that intentional? I don't think the generated cases are exhaustive.

Reply:
I don't think I was the original author, but we definitely want this to be exhaustive. We could do that in a separate PR, or here. If you can come up with a failing test in addition that would be amazing.

Reply:
I'm ok adding it + test here :+1: